src/move/emojicoin_dot_fun/sources/tests.move

```diff
@@ -94,6 +142,170 @@
+        periodic_state_trackers: vector<TestPeriodicStateTracker>,
+        aptos_coin_balance: u64,
+        emojicoin_balance: u64,
+        emojicoin_lp_balance: u64,
+    }
+
+    struct TestPeriodicState has copy, drop, store {
+        market_metadata: TestMarketMetadata,
+        periodic_state_metadata: TestPeriodicStateMetadata,
+        open_price_q64: u128,
+        high_price_q64: u128,
+        low_price_q64: u128,
+        close_price_q64: u128,
+        volume_base: u128,
+        volume_quote: u128,
+        integrator_fees: u128,
+        pool_fees_base: u128,
+        pool_fees_quote: u128,
+        n_swaps: u64,
+        n_chat_messages: u64,
+        starts_in_bonding_curve: bool,
+        ends_in_bonding_curve: bool,
+        tvl_per_lp_coin_growth_q64: u128,
+    }
+
+    struct TestPeriodicStateTracker has copy, drop, store {
+        start_time: u64,
+        period: u64,
+        open_price_q64: u128,
+        high_price_q64: u128,
+        low_price_q64: u128,
+        close_price_q64: u128,
+        volume_base: u128,
+        volume_quote: u128,
+        integrator_fees: u128,
+        pool_fees_base: u128,
+        pool_fees_quote: u128,
+        n_swaps: u64,
+        n_chat_messages: u64,
+        starts_in_bonding_curve: bool,
+        ends_in_bonding_curve: bool,
+        tvl_to_lp_coin_ratio_start: TestTVLtoLPCoinRatio,
+        tvl_to_lp_coin_ratio_end: TestTVLtoLPCoinRatio,
+    }
+
+    struct TestPeriodicStateMetadata has copy, drop, store {
+        start_time: u64,
+        period: u64,
+        emit_time: u64,
+        emit_market_nonce: u64,
+        trigger: u8,
+    }
+
+    struct TestRegistryView has copy, drop, store {
+        registry_address: address,
+        nonce: u64,
+        last_bump_time: u64,
+        n_markets: u64,
+        cumulative_quote_volume: u128,
+        total_quote_locked: u128,
+        total_value_locked: u128,
+        market_cap: u128,
+        fully_diluted_value: u128,
+        cumulative_integrator_fees: u128,
+        cumulative_swaps: u64,
+        cumulative_chat_messages: u64,
+    }
+
+    struct TestTVLtoLPCoinRatio has copy, drop, store {
+        tvl: u128,
+        lp_coins: u128,
+    }
+
+    struct TestReserves has copy, drop, store {
+        base: u64,
+        quote: u64,
+    }
+
+    struct TestSequenceInfo has copy, drop, store {
+        nonce: u64,
+        last_bump_time: u64,
+    }
+
+    struct TestState has copy, drop, store {
+        market_metadata: TestMarketMetadata,
+        state_metadata: TestStateMetadata,
+        clamm_virtual_reserves: TestReserves,
+        cpamm_real_reserves: TestReserves,
+        lp_coin_supply: u128,
+        cumulative_stats: TestCumulativeStats,
+        instantaneous_stats: TestInstantaneousStats,
+        last_swap: TestLastSwap,
+    }
+
+    struct TestStateMetadata has copy, drop, store {
+        market_nonce: u64,
+        bump_time: u64,
+        trigger: u8,
+    }
+
```

Comment:
What's the reasoning behind duplicating all these Test structs? Is there a dependency issue or something? Just wondering because duplication of this logic is probably error prone, but I assume the compiler is just being a pain again..? 😓

Reply:
I've chosen to use struct here for mocking so that we can enforce named fields

When testing Econia v4, it ended up that I had to write a bunch of assert statements for each individual value that was changed

As a result there ended up being a ton of copy-paste, which is not so maintainable when one thing changes and you have to go through and update every test

Hence to abstract everything there is now a suite of `base_<struct_name>` functions, which return default values

Then during a test you only need to manually modify certain fields that deviate from the base, and call `assert_<struct_name>` which does all the unpacking internally

Check out `register_market_comprehensive_state_assertion()`, which would be way more lines if we were manually asserting everything and not relying on named mock fields

Reply:
Yes this is correct

I've updated other followup testing task ECO-1640 to update to `Mock` and also add some docs

Will merge this so APIs are available in `main`, then address those two items in next testing PR